// Copyright 2021 The gRPC Authors

syntax = "proto3";

package envoy.service.status.v3;

import "envoy/admin/v3/config_dump_shared.proto";
import "envoy/config/core/v3/base.proto";

import "google/protobuf/any.proto";
import "google/protobuf/timestamp.proto";

service ClientStatusDiscoveryService {
  rpc StreamClientStatus(stream ClientStatusRequest) returns (stream ClientStatusResponse) {}
  rpc FetchClientStatus(ClientStatusRequest) returns (ClientStatusResponse) {}
}

enum ConfigStatus {

  UNKNOWN = 0;

  SYNCED = 1;

  NOT_SENT = 2;

  STALE = 3;

  ERROR = 4;
}

message ClientStatusRequest {

  config.core.v3.Node node = 2;
}

message PerXdsConfig {

  ConfigStatus status = 1;

  oneof per_xds_config {
    admin.v3.ListenersConfigDump listener_config = 2;

    admin.v3.ClustersConfigDump cluster_config = 3;

    admin.v3.RoutesConfigDump route_config = 4;

    admin.v3.EndpointsConfigDump endpoint_config = 6;
  }
}

message ClientConfig {

  message GenericXdsConfig {

    string type_url = 1;

    string name = 2;

    string version_info = 3;

    google.protobuf.Any xds_config = 4;

    google.protobuf.Timestamp last_updated = 5;

    ConfigStatus config_status = 6;

    admin.v3.ClientResourceStatus client_status = 7;

    admin.v3.UpdateFailureState error_state = 8;

    bool is_static_resource = 9;
  }

  config.core.v3.Node node = 1;

  repeated PerXdsConfig xds_config = 2 [deprecated = true];

  repeated GenericXdsConfig generic_xds_configs = 3;

  string client_scope = 4;
}

message ClientStatusResponse {

  repeated ClientConfig config = 1;
}
